File: gherkin/java/src/main/java/io/cucumber/gherkin/GherkinParser.java
package io.cucumber.gherkin;

import io.cucumber.gherkin.ParserException.CompositeParserException;
import io.cucumber.messages.IdGenerator;
import io.cucumber.messages.types.Envelope;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.ParseError;
import io.cucumber.messages.types.Source;
import io.cucumber.messages.types.SourceReference;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Stream;

import static io.cucumber.gherkin.EncodingParser.readWithEncodingFromSource;
import static io.cucumber.messages.types.SourceMediaType.TEXT_X_CUCUMBER_GHERKIN_PLAIN;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toCollection;

/**
 * Main entry point for the Gherkin library
 */
public final class GherkinParser {

    public final static class Builder {
        private boolean includeSource = true;
        private boolean includeGherkinDocument = true;
        private boolean includePickles = true;
        private IdGenerator idGenerator = () -> UUID.randomUUID().toString();

        private Builder() {

        }

        public Builder includeSource(boolean includeSource) {
            this.includeSource = includeSource;
            return this;
        }

        public Builder includeGherkinDocument(boolean includeGherkinDocument) {
            this.includeGherkinDocument = includeGherkinDocument;
            return this;
        }

        public Builder includePickles(boolean includePickles) {
            this.includePickles = includePickles;
            return this;
        }

        public Builder idGenerator(IdGenerator idGenerator) {
            this.idGenerator = requireNonNull(idGenerator);
            return this;
        }

        public GherkinParser build() {
            return new GherkinParser(includeSource, includeGherkinDocument, includePickles, idGenerator);
        }

    }

    private final boolean includeSource;
    private final boolean includeGherkinDocument;
    private final boolean includePickles;
    private final IdGenerator idGenerator;
    private final PickleCompiler pickleCompiler;

    private GherkinParser(boolean includeSource, boolean includeGherkinDocument, boolean includePickles,
            IdGenerator idGenerator) {
        this.includeSource = includeSource;
        this.includeGherkinDocument = includeGherkinDocument;
        this.includePickles = includePickles;
        this.idGenerator = requireNonNull(idGenerator);
        this.pickleCompiler = new PickleCompiler(idGenerator);
    }

    public static Builder builder() {
        return new Builder();
    }

    public Stream<Envelope> parse(Path path) throws IOException {
        // .feature.md files are not supported
        return parse(path.toUri().toString(), Files.readAllBytes(path));
    }

    public Stream<Envelope> parse(String uri, InputStream source) throws IOException {
        return parse(uri, InputStreams.readAllBytes(source));
    }

    public Stream<Envelope> parse(String uri, byte[] source) {
        String withEncodingFromSource = readWithEncodingFromSource(source);
        return parse(Envelope.of(new Source(uri, withEncodingFromSource, TEXT_X_CUCUMBER_GHERKIN_PLAIN)));
    }

    public Stream<Envelope> parse(Envelope envelope) {
        requireNonNull(envelope);

        List<Envelope> messages = new ArrayList<>();

        if (includeSource) {
            messages.add(envelope);
        }

        envelope.getSource()
                .map(this::parse)
                .ifPresent(messages::addAll);

        return messages.stream();
    }

    private List<Envelope> parse(Source source) {
        return parse(source.getUri(), source.getData());
    }

    private List<Envelope> parse(String uri, String data) {
        List<Envelope> messages = new ArrayList<>();
        GherkinDocumentBuilder documentBuilder = new GherkinDocumentBuilder(idGenerator, uri);
        Parser<GherkinDocument> parser = new Parser<>(documentBuilder);

        try {
            GherkinDocument gherkinDocument = parser.parse(data, uri);
            if (includeGherkinDocument) {
                messages.add(Envelope.of(gherkinDocument));
            }
            if (includePickles) {
                pickleCompiler.compile(gherkinDocument, uri)
                        .stream()
                        .map(Envelope::of)
                        .collect(toCollection(() -> messages));
            }
        } catch (CompositeParserException composite) {
            composite.errors.stream()
                    .map(error -> createParseError(error, uri))
                    .collect(toCollection(() -> messages));
        } catch (ParserException error) {
            messages.add(createParseError(error, uri));
        }
        return messages;
    }

    private Envelope createParseError(ParserException e, String uri) {
        long line = e.location.getLine();
        long column = e.location.getColumn();
        return Envelope.of(new ParseError(
                new SourceReference(
                        uri,
                        null,
                        null,
                        // We want 0 values not to be serialised, which is why we set them to null
                        // This is a legacy requirement brought over from old protobuf behaviour
                        new io.cucumber.messages.types.Location(
                                line,
                                column == 0 ? null : column
                        )
                ),
                e.getMessage()
        ));
    }

}

File: gherkin/php/src/GherkinParser.php
<?php

declare(strict_types=1);

namespace Cucumber\Gherkin;

use Cucumber\Gherkin\ParserException\CompositeParserException;
use Cucumber\Messages\Envelope;
use Cucumber\Messages\GherkinDocument;
use Cucumber\Messages\Id\IdGenerator;
use Cucumber\Messages\Id\IncrementingIdGenerator;
use Cucumber\Messages\Id\UuidIdGenerator;
use Cucumber\Messages\Location as MessageLocation;
use Cucumber\Messages\ParseError;
use Cucumber\Messages\Source;
use Cucumber\Messages\SourceReference;
use Generator;

/**
 * Parses a Gherkin document (or list of Source envelopes) and emits Cucumber Messages envelopes
 */
final class GherkinParser
{
    private readonly PickleCompiler $pickleCompiler;
    private readonly IdGenerator $idGenerator;

    /**
     * @param bool $predictableIds Ignored if IdGenerator is provided
     */
    public function __construct(
        private readonly bool $predictableIds = false,
        private readonly bool $includeSource = true,
        private readonly bool $includeGherkinDocument = true,
        private readonly bool $includePickles = true,
        ?IdGenerator $idGenerator = null,
    ) {
        $this->idGenerator = $idGenerator ?? ($this->predictableIds ? new IncrementingIdGenerator() : new UuidIdGenerator());
        $this->pickleCompiler = new PickleCompiler($this->idGenerator);
    }

    /**
     * Parse a single string of Gherkin (with its URI)
     *
     * @return Generator<Envelope>
     */
    public function parseString(string $uri, string $data): Generator
    {
        yield from $this->parse([
            new Source(uri: $uri, data: $data),
        ]);
    }

    /**
     * @param iterable<Source> $sources
     *
     * @return Generator<Envelope>
     */
    public function parse(iterable $sources): Generator
    {
        foreach ($sources as $source) {
            if (!$source instanceof Source) {
                throw new \InvalidArgumentException('Can only parse Source objects');
            }

            if ($this->includeSource) {
                yield new Envelope(source: $source);
            }

            try {
                $gherkinDocument = $this->parseGherkinDocument($source);

                if ($this->includeGherkinDocument) {
                    yield new Envelope(gherkinDocument: $gherkinDocument);
                }

                if ($this->includePickles) {
                    foreach ($this->pickleCompiler->compile($gherkinDocument, $source->uri) as $pickle) {
                        yield new Envelope(pickle: $pickle);
                    }
                }
            } catch (CompositeParserException $composite) {
                foreach ($composite->errors as $error) {
                    yield $this->createParseError($error, $source->uri);
                }
            } catch (ParserException $error) {
                yield $this->createParseError($error, $source->uri);
            }
        }
    }

    private function parseGherkinDocument(Source $source): GherkinDocument
    {
        $builder = new GherkinDocumentBuilder($source->uri, $this->idGenerator);
        $parser = new Parser($builder);

        return $parser->parse(
            $source->uri,
            new StringTokenScanner($source->data),
            new TokenMatcher(),
        );
    }

    private function createParseError(ParserException $error, string $uri): Envelope
    {
        $ref = new SourceReference(
            uri: $uri,
            location: new MessageLocation(
                line: $error->location->line,
                column: $error->location->column === 0 ? null : $error->location->column,
            ),
        );

        return new Envelope(parseError: new ParseError($ref, $error->getMessage()));
    }
}
